tracing-subscriber/src/registry/fmt.rs

```diff
@@ -0,0 +1,315 @@
+        Self {
+            is_interested: Box::new(|_| true),
+            fmt_fields: format::DefaultFields::default(),
+            fmt_event: format::Format::default(),
+            make_writer: io::stdout,
+            _inner: PhantomData,
+        }
+    }
+}
+
+impl<S, N, E, W> FmtLayer<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    #[inline]
+    fn make_ctx<'a>(&'a self, ctx: Context<'a, S>) -> FmtContext<'a, S, N> {
+        FmtContext {
+            ctx,
+            fmt_fields: &self.fmt_fields,
+        }
+    }
+}
+
+// === impl Formatter ===
+
+impl<S, N, E, W> Layer<S> for FmtLayer<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    fn new_span(&self, attrs: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let span = span.data;
+        let mut fields = String::new();
+        if self.fmt_fields.format_fields(&mut fields, attrs).is_ok() {
+            span.extensions_mut().insert(fields);
+        }
+    }
+
+    fn on_record(&self, id: &Id, values: &Record<'_>, ctx: Context<'_, S>) {
+        if let Some(span) = ctx.span(id) {
+            let mut buf = String::new();
+            if self.fmt_fields.format_fields(&mut buf, values).is_ok() {
+                let mut writer = self.make_writer.make_writer();
+                let _ = io::Write::write_all(&mut writer, buf.as_bytes());
+            }
+        }
+    }
+
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        thread_local! {
+            static BUF: RefCell<String> = RefCell::new(String::new());
+        }
+
+        BUF.with(|buf| {
+            let borrow = buf.try_borrow_mut();
+            let mut a;
+            let mut b;
+            let mut buf = match borrow {
+                Ok(buf) => {
+                    a = buf;
+                    &mut *a
+                }
+                _ => {
+                    b = String::new();
+                    &mut b
+                }
+            };
+
+            if (self.is_interested)(event) {
+                let ctx = self.make_ctx(ctx);
+                if self.fmt_event.format_event(&ctx, &mut buf, event).is_ok() {
+                    let mut writer = self.make_writer.make_writer();
+                    let _ = io::Write::write_all(&mut writer, buf.as_bytes());
+                }
+            }
+
+            buf.clear();
+        });
+    }
+}
+
+/// `FmtContext` is used to propogate subscriber context to tracing_subscriber::fmt.
+pub struct FmtContext<'a, S, N>
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
```

Comment:
Generally, I prefer not to put trait bounds on struct definitions except when they're _required_ — if we got rid of these, the `fmt::Debug` impl wouldn't need to be bounded either...

Reply:
fair; removed!